lang/src/env_access.rs

```diff
@@ -0,0 +1,359 @@
+        env::invoke_runtime::<T>(params)
+    }
+
+    /// Invokes a contract message.
+    ///
+    /// # Note
+    ///
+    /// This is a low level way to invoke another smart contract.
+    /// Prefer to use the ink! guided and type safe approach to using this.
+    ///
+    /// # Errors
+    ///
+    /// - If the called contract does not exist.
+    /// - If the called contract is a tombstone.
+    /// - If arguments passed to the called contract message are invalid.
+    /// - If the called contract execution has trapped.
+    /// - If the called contract ran out of gas upon execution.
+    /// - If given too few endowment.
+    pub fn invoke_contract(self, params: &CallParams<T, ()>) -> Result<()> {
+        env::invoke_contract::<T>(params)
+    }
+
+    /// Evaluates a contract message and returns its result.
+    ///
+    /// # Note
+    ///
+    /// This is a low level way to evaluate another smart contract.
+    /// Prefer to use the ink! guided and type safe approach to using this.
+    ///
+    /// # Errors
+    ///
+    /// - If the called contract does not exist.
+    /// - If the called contract is a tombstone.
+    /// - If arguments passed to the called contract message are invalid.
+    /// - If the called contract execution has trapped.
+    /// - If the called contract ran out of gas upon execution.
+    /// - If given too few endowment.
+    /// - If the returned value failed to decode properly.
+    pub fn eval_contract<R>(self, params: &CallParams<T, ReturnType<R>>) -> Result<R>
+    where
+        R: scale::Decode,
+    {
+        env::eval_contract::<T, R>(params)
+    }
+
+    /// Instantiates another contract.
+    ///
+    /// # Note
+    ///
+    /// This is a low level way to instantiate another smart contract.
+    /// Prefer to use the ink! guided and type safe approach to using this.
+    ///
+    /// # Errors
+    ///
+    /// - If the code hash is invalid.
+    /// - If the arguments passed to the instantiation process are invalid.
+    /// - If the instantiation process traps.
+    /// - If the instantiation process runs out of gas.
+    /// - If given too few endowment.
+    /// - If the returned account ID failed to decode properly.
+    pub fn create_contract<C>(
+        self,
+        params: &InstantiateParams<T, C>,
+    ) -> Result<T::AccountId> {
+        env::create_contract::<T, C>(params)
+    }
+
+    /// Restores a smart contract in tombstone state.
+    ///
+    /// # Params
+    ///
+    /// - `account_id`: Account ID of the to-be-restored contract.
+    /// - `code_hash`: Code hash of the to-be-restored contract.
+    /// - `rent_allowance`: Rent allowance of the restored contract
+    ///                     upon successful restoration.
+    /// - `filtered_keys`: Storage keys that will be removed for the tombstone hash
+    ///                    match calculation that decide whether the original contract
+    ///                    storage and the storage of the restorer contract is equal.
```

Comment:
Sentence is hard to read, perhaps something like: `Storage keys to be excluded when calculating the tombstone hash, which is used to decide whether the original contract and the to-be-restored contract have matching storage.`

Reply:
yeah much better, thanks!